near-sdk/src/store/tree_map/mod.rs

```diff
@@ -0,0 +1,2118 @@
+                match node.lft {
+                    Some(lft) => at = lft,
+                    None => break,
+                }
+            }
+        }
+        seen
+    }
+
+    fn equal_at<Q>(&self, mut at: FreeListIndex, key: &Q) -> Option<&K>
+    where
+        K: Borrow<Q>,
+        Q: ?Sized + Ord,
+    {
+        while let Some(node) = self.node(at) {
+            let k: &Q = node.key.borrow();
+            if k.eq(key) {
+                return Some(&node.key);
+            } else if k.lt(key) {
+                match node.rgt {
+                    Some(rgt) => at = rgt,
+                    None => break,
+                }
+            } else {
+                match node.lft {
+                    Some(lft) => at = lft,
+                    None => break,
+                }
+            }
+        }
+        None
+    }
+
+    /// Returns (node, parent node) for a node that holds the `key`.
+    /// For root node, same node is returned for node and parent node.
+    #[allow(clippy::type_complexity)]
+    fn lookup_at<Q: ?Sized>(
+        &self,
+        mut at: FreeListIndex,
+        key: &Q,
+    ) -> Option<(NodeAndIndex<K>, Option<(FreeListIndex, &Node<K>, Edge)>)>
+    where
+        K: Borrow<Q>,
+        Q: BorshSerialize + Eq + PartialOrd,
+    {
+        let mut p = None;
+        let mut curr = Some(expect(self.node(at)));
+        while let Some(node) = curr {
+            let node_key: &Q = node.key.borrow();
+            if node_key.eq(key) {
+                return Some(((at, node), p));
+            } else if node_key.lt(key) {
+                match node.rgt {
+                    Some(rgt) => {
+                        p = Some((at, node, Edge::Right));
+                        at = rgt;
+                    }
+                    None => break,
+                }
+            } else {
+                match node.lft {
+                    Some(lft) => {
+                        p = Some((at, node, Edge::Left));
+                        at = lft;
+                    }
+                    None => break,
+                }
+            }
+            curr = self.node(at);
+        }
+        None
+    }
+}
+
+impl<K> Tree<K>
+where
+    K: Ord + BorshSerialize + BorshDeserialize + Clone,
+{
+    fn internal_insert(&mut self, key: K) {
+        if let Some(root) = self.root {
+            let node = expect(self.node(root)).clone();
+            self.root = Some(self.insert_at(node, root, key));
+        } else {
+            self.root = Some(self.nodes.insert(Node::of(key)));
+        }
+    }
+
+    fn insert_at(&mut self, mut node: Node<K>, id: FreeListIndex, key: K) -> FreeListIndex {
+        if key.eq(&node.key) {
+            // This branch should not be hit, because we check for existence in insert.
+            id
+        } else {
+            if key.lt(&node.key) {
+                let idx = match node.lft {
+                    Some(lft) => self.insert_at(expect(self.node(lft)).clone(), lft, key),
```

Comment:
Have you explored whether it is possible to get rid of all these clones by writing another auxiliary method like this one:
```rust
fn node_mut(&mut self, id: FreeListIndex) -> Option<&mut Node<K>> {
    self.nodes.get_mut(id)
}
```

And then just operating with `&mut Node<K>` in this method (and maybe some others too)?

Reply:
yeah, absolutely did, the issue with the current impl is you can't borrow this mutably and borrow other parts of the vec immutably. Yes we know there isn't overlap, but would need unsafe for this to happen. This is why clones currently happen in the implementation, but is still an improvement over the existing implementation because it was loading and deserializing the value where the clone was.

This can be optimized in the future without breaking changes with the current API.

Reply:
I see. I thought something like that might be the case, thanks for confirming.